scotty/pitch_angle_analysis.py

```diff
@@ -0,0 +1,858 @@
+            title_label = "dist"
+            variable = "cutoff_distance"
+        else:
+            raise ValueError("Provided measure is not 'rho' or 'm'.")
+
+        if const_angle not in self.get_Coordinate_array(const_angle_str):
+            raise ValueError("Value of const_angle not in original coordinate array.")
+
+        coords_dict = {const_angle_str: const_angle}
+        xdimension = "frequency"
+        ydimension = var_angle_str
+        spline_surface = self.create_2Dspline(
+            variable, xdimension, ydimension, coords_dict
+        )
+
+        x_coordinates = self.get_Coordinate_array(xdimension)
+        y_coordinates = self.get_Coordinate_array(ydimension)
+
+        if bounds == None:
+            x_start = x_coordinates[0]
+            x_stop = x_coordinates[-1]
+            y_start = y_coordinates[0]
+            y_stop = y_coordinates[-1]
+        else:
+            x_start = bounds[0]
+            x_stop = bounds[1]
+            y_start = bounds[2]
+            y_stop = bounds[3]
+
+        x_array = np.linspace(x_start, x_stop, num=30)
+        y_array = np.linspace(y_start, y_stop, num=30)
+        Xgrid, Ygrid = np.meshgrid(x_array, y_array, indexing="ij")
+        Zgrid = spline_surface(x_array, y_array)
+
+        # Convert from poloidal flux to rho
+        if measure == "rho":
+            Zgrid = np.sqrt(Zgrid)
+
+        # Interpolate and apply the reflection mask to remove inaccessible regions
+        reflection_mask = (
+            self.dataset["reflection_mask"]
+            .loc[coords_dict]
+            .transpose(xdimension, ydimension)
+            .values
+        )
+        mask_function = RegularGridInterpolator(
+            (x_coordinates, y_coordinates),
+            reflection_mask,
+            method="nearest",
+            bounds_error=False,
+        )
+        interp_mask = mask_function((Xgrid, Ygrid))
+        Z_masked = np.ma.array(Zgrid, mask=interp_mask)
+
+        fig, ax = plt.subplots()
+        if mask_flag:
+            CS = ax.contour(
+                Xgrid, Ygrid, Z_masked, levels=20, cmap="plasma_r", corner_mask=True
+            )
+        else:
+            CS = ax.contour(
+                Xgrid, Ygrid, Zgrid, levels=20, cmap="plasma_r", corner_mask=True
+            )
+
+        fig.suptitle(f"Cutoff {title_label}, {const_angle_str}={const_angle}$^\circ$")
+        ax.set_xlabel("frequency/GHz")
+        ax.set_ylabel(f"{var_angle_str}/$^\circ$")
+        ax.clabel(CS, inline=True, fontsize=6)
+        fig.colorbar(CS)
+        if save_path:
+            plt.savefig(
+                save_path + f"cutoff_contour_{const_angle_str}{const_angle}.jpg",
+                dpi=200,
+            )
+        return (Xgrid, Ygrid, Zgrid, interp_mask)
+
+    # To be implemented
+    """ 
+    def plot_cutoff_angles(self, poloidal_angle):
+        return
+
+    def plot_opt_tor_vs_freq(self, poloidal_angle):
+        return
+    """
+
+
+class SweepFromNC(SweepDataset):
```

Comment:
Another way you might like to implement this is in `SweepDataset` as a [`classmethod`](https://docs.python.org/3/library/functions.html#classmethod), for example:

```python
class SweepDataset:
    def __init__(self, ds):
        self.dataset = ds

    @classmethod
    def from_netcdf(cls, path_string):
        ...
        return cls(ds)

    @classmethod
    def from_scan(cls, ...):
        # existing __init__ implementation
```

This pattern is a pretty common way to have multiple constructors for Python classes. You might prefer the existing way, of course, and that's also fine.